presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SingleDistinctOptimizer.java

```diff
@@ -0,0 +1,150 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.optimizations;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.sql.planner.PlanNodeIdAllocator;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.SymbolAllocator;
+import com.facebook.presto.sql.planner.plan.AggregationNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.MarkDistinctNode;
+import com.facebook.presto.sql.planner.plan.PlanRewriter;
+import com.facebook.presto.sql.planner.plan.ProjectNode;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.facebook.presto.sql.tree.NullLiteral;
+import com.facebook.presto.sql.tree.QualifiedNameReference;
+import com.facebook.presto.spi.type.Type;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.facebook.presto.sql.planner.optimizations.HashGenerationOptimizer.getHashExpression;
+import static com.facebook.presto.sql.planner.plan.AggregationNode.Step.SINGLE;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * Converts Single Distinct Aggregation into GroupBy
+ *
+ * Rewrite if and only if
+ *  1 all aggregation functions have a single common distinct mask symbol
+ *  2 all aggregation functions have mask
+ *
+ * Rewrite MarkDistinctNode into AggregationNode(use DistinctSymbols as GroupBy)
+ * Add ProjectNode on top of the new AggregationNode, which adds null assignment for mask
+ * All unused mask will be removed by PruneUnreferencedOutputs
+ * Remove Distincts in the original AggregationNode
+ */
+public class SingleDistinctOptimizer
+        extends PlanOptimizer
+{
+    @Override
+    public PlanNode optimize(PlanNode plan, Session session, Map<Symbol, Type> types, SymbolAllocator symbolAllocator, PlanNodeIdAllocator idAllocator)
+    {
+        return PlanRewriter.rewriteWith(new Optimizer(idAllocator, symbolAllocator), plan, Optional.empty());
+    }
+
+    private static class Optimizer
+            extends PlanRewriter<Optional<Symbol>>
+    {
+        private final SymbolAllocator symbolAllocator;
+        private final PlanNodeIdAllocator idAllocator;
+
+        private Optimizer(PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator)
+        {
+            this.idAllocator = checkNotNull(idAllocator, "idAllocator is null");
+            this.symbolAllocator = checkNotNull(symbolAllocator, "symbolAllocator is null");
+        }
+
+        @Override
+        public PlanNode visitAggregation(AggregationNode node, RewriteContext<Optional<Symbol>> context)
+        {
+            if (!canOptimize(node)) {
```

Comment:
I'd inline the canOptimize method here. Otherwise, it's not very clear that the line below that extracts the mask is working off of the assumption that there's a single mask.

So, possibly, rewrite it as:

```java
Set<Symbol> masks = ImmutableSet.copyOf(node.getMasks().values());

if (masks.size() > 1 || node.getMasks().size() != node.getAggregations.size()) {
    return node;
}

Symbol mask = Iterables.getOnlyElement(masks);
...
```